#Some basic processing for sequences that have been generated by Sanger sequencing and then trimmed by Chromas.


san.names = list.files("original_data/fasta/sanger.sequencing/")
sanger.sequences = paste("original_data/fasta/sanger.sequencing/", list.files("original_data/fasta/sanger.sequencing/"), sep = "")
sanger.sequences1 = lapply(sanger.sequences, readDNAStringSet)
san2 = DNAStringSet()
for(i in sanger.sequences1){
    san2 = c(san2, i)
}


names(san2) = san.names

#reverse sequences with rev. primers
san2[grep("R", names(san2))] = reverseComplement(san2[grep("R", names(san2))])

#remove t.700 sequences
san2 = san2[-grep("700", names(san2))]


g = strsplit(names(san2), "_")

orig.order = unlist(lapply(g, function(x) as.numeric(x[[1]])))
order.new = sort(orig.order)

san2 = san2[match(order.new, orig.order)]


writeXStringSet(san2, "original_data/fasta/sanger.sequencing/all.sanger.fa")
