all.coverage.files = c("Sample_10-23_1190092/bams/a8.cov", "Sample_11-24_1190110/bams/a8.cov", "Sample_12-25_1190126/bams/a8.cov", "Sample_1-3_1190034/bams/a8.cov", "Sample_13-26_1190128/bams/a8.cov", "Sample_14-27_1190139/bams/a8.cov", "Sample_15-28_1190145/bams/a8.cov", "Sample_16-29_1190149/bams/a8.cov", "Sample_17-30_1190160/bams/a8.cov", "Sample_18-31_1190166/bams/a8.cov", "Sample_19-32_1190181/bams/a8.cov", "Sample_1a-1_1190141/bams/a8.cov", "Sample_20-33_1190199/bams/a8.cov", "Sample_21-34_1190209/bams/a8.cov", "Sample_22-35_1190216/bams/a8.cov", "Sample_23-36_1190218/bams/a8.cov", "Sample_24-37_1190219/bams/a8.cov", "Sample_25-38_1190224/bams/a8.cov", "Sample_26-39_1190224/bams/a8.cov", "Sample_27-40_1190231/bams/a8.cov", "Sample_2-8_1190299/bams/a8.cov", "Sample_28-41_1190239/bams/a8.cov", "Sample_29-42_1190246/bams/a8.cov", "Sample_2a-2_1190292/bams/a8.cov", "Sample_30-43_1190254/bams/a8.cov", "Sample_3-14_1190007/bams/a8.cov", "Sample_31-44_1190264/bams/a8.cov", "Sample_32-45_1190273/bams/a8.cov", "Sample_33-46_1190281/bams/a8.cov", "Sample_34-47_1190291/bams/a8.cov", "Sample_35-48_1190300/bams/a8.cov", "Sample_36-50_1190313/bams/a8.cov", "Sample_37-51_1190324/bams/a8.cov", "Sample_38-52_1190326/bams/a8.cov", "Sample_39-53_1190349/bams/a8.cov", "Sample_3a-4_1190238/bams/a8.cov", "Sample_40-54_1190352/bams/a8.cov", "Sample_41-55_1190355/bams/a8.cov", "Sample_4-16_1190032/bams/a8.cov", "Sample_42-56_1190360/bams/a8.cov", "Sample_43-57_1190387/bams/a8.cov", "Sample_44-58_1190396/bams/a8.cov", "Sample_45-59_1190398/bams/a8.cov", "Sample_46-60_1190406/bams/a8.cov", "Sample_47-61_1190420/bams/a8.cov", "Sample_48-62_1190433/bams/a8.cov", "Sample_49-63_1190440/bams/a8.cov", "Sample_4a-5_1190308/bams/a8.cov", "Sample_50-65_1190451/bams/a8.cov", "Sample_51-66_1190460/bams/a8.cov", "Sample_5-17_1190040/bams/a8.cov", "Sample_52-67_1190468/bams/a8.cov", "Sample_53-68_1190471/bams/a8.cov", "Sample_54-69_1190474/bams/a8.cov", "Sample_55-70_1190475/bams/a8.cov", "Sample_56-71_1190481/bams/a8.cov", "Sample_57-72_1190483/bams/a8.cov", "Sample_58-73_1190496/bams/a8.cov", "Sample_59-74_1190507/bams/a8.cov", "Sample_5a-6_1190103/bams/a8.cov", "Sample_60-75_1190546/bams/a8.cov", "Sample_61-76_1190551/bams/a8.cov", "Sample_6-18_1190042/bams/a8.cov", "Sample_62-77_1190560/bams/a8.cov", "Sample_63-78_1190562/bams/a8.cov", "Sample_64-81_1190579/bams/a8.cov", "Sample_65-82_1190580/bams/a8.cov", "Sample_66-83_1190591/bams/a8.cov", "Sample_67-85_1190624/bams/a8.cov", "Sample_68-86_1190629/bams/a8.cov", "Sample_69-87_1190637/bams/a8.cov", "Sample_6a-7_1190827/bams/a8.cov", "Sample_70-88_1190639/bams/a8.cov", "Sample_71-89_1190651/bams/a8.cov", "Sample_7-19_1190044/bams/a8.cov", "Sample_72-90_1190652/bams/a8.cov", "Sample_73-91_1190662/bams/a8.cov", "Sample_74-92_1190670/bams/a8.cov", "Sample_75-93_1190671/bams/a8.cov", "Sample_76-95_1190683/bams/a8.cov", "Sample_77-96_1190685/bams/a8.cov", "Sample_78-97_1190690/bams/a8.cov", "Sample_79-98_1190694/bams/a8.cov", "Sample_7a-9_1190627/bams/a8.cov", "Sample_80-99_1190698/bams/a8.cov", "Sample_81-100_1190700/bams/a8.cov", "Sample_8-20_1190045/bams/a8.cov", "Sample_82-101_1190704/bams/a8.cov", "Sample_83-102_1190705/bams/a8.cov", "Sample_84-103_1190707/bams/a8.cov", "Sample_85-104_1190722/bams/a8.cov", "Sample_86-105_1190731/bams/a8.cov", "Sample_87-107_1190740/bams/a8.cov", "Sample_88-108_1190742/bams/a8.cov", "Sample_89-109_1190746/bams/a8.cov", "Sample_8a-10_1190811/bams/a8.cov", "Sample_90-110_1190747/bams/a8.cov", "Sample_91-111_1190749/bams/a8.cov", "Sample_9-21_1190079/bams/a8.cov", "Sample_92-112_1190753/bams/a8.cov", "Sample_93-113_1190771/bams/a8.cov", "Sample_94-114_1190777/bams/a8.cov", "Sample_95-115_1190784/bams/a8.cov", "Sample_96-116_1190788/bams/a8.cov")
other.coverage.files = c("Sample_11-24_1190110/bams/a8.cov", "Sample_12-25_1190126/bams/a8.cov", "Sample_1-3_1190034/bams/a8.cov", "Sample_13-26_1190128/bams/a8.cov", "Sample_14-27_1190139/bams/a8.cov", "Sample_15-28_1190145/bams/a8.cov", "Sample_16-29_1190149/bams/a8.cov", "Sample_17-30_1190160/bams/a8.cov", "Sample_18-31_1190166/bams/a8.cov", "Sample_19-32_1190181/bams/a8.cov", "Sample_1a-1_1190141/bams/a8.cov", "Sample_20-33_1190199/bams/a8.cov", "Sample_21-34_1190209/bams/a8.cov", "Sample_22-35_1190216/bams/a8.cov", "Sample_23-36_1190218/bams/a8.cov", "Sample_24-37_1190219/bams/a8.cov", "Sample_25-38_1190224/bams/a8.cov", "Sample_26-39_1190224/bams/a8.cov", "Sample_27-40_1190231/bams/a8.cov", "Sample_2-8_1190299/bams/a8.cov", "Sample_28-41_1190239/bams/a8.cov", "Sample_29-42_1190246/bams/a8.cov", "Sample_2a-2_1190292/bams/a8.cov", "Sample_30-43_1190254/bams/a8.cov", "Sample_3-14_1190007/bams/a8.cov", "Sample_31-44_1190264/bams/a8.cov", "Sample_32-45_1190273/bams/a8.cov", "Sample_33-46_1190281/bams/a8.cov", "Sample_34-47_1190291/bams/a8.cov", "Sample_35-48_1190300/bams/a8.cov", "Sample_36-50_1190313/bams/a8.cov", "Sample_37-51_1190324/bams/a8.cov", "Sample_38-52_1190326/bams/a8.cov", "Sample_39-53_1190349/bams/a8.cov", "Sample_3a-4_1190238/bams/a8.cov", "Sample_40-54_1190352/bams/a8.cov", "Sample_41-55_1190355/bams/a8.cov", "Sample_4-16_1190032/bams/a8.cov", "Sample_42-56_1190360/bams/a8.cov", "Sample_43-57_1190387/bams/a8.cov", "Sample_44-58_1190396/bams/a8.cov", "Sample_45-59_1190398/bams/a8.cov", "Sample_46-60_1190406/bams/a8.cov", "Sample_47-61_1190420/bams/a8.cov", "Sample_48-62_1190433/bams/a8.cov", "Sample_49-63_1190440/bams/a8.cov", "Sample_4a-5_1190308/bams/a8.cov", "Sample_50-65_1190451/bams/a8.cov", "Sample_51-66_1190460/bams/a8.cov", "Sample_5-17_1190040/bams/a8.cov", "Sample_52-67_1190468/bams/a8.cov", "Sample_53-68_1190471/bams/a8.cov", "Sample_54-69_1190474/bams/a8.cov", "Sample_55-70_1190475/bams/a8.cov", "Sample_56-71_1190481/bams/a8.cov", "Sample_57-72_1190483/bams/a8.cov", "Sample_58-73_1190496/bams/a8.cov", "Sample_59-74_1190507/bams/a8.cov", "Sample_5a-6_1190103/bams/a8.cov", "Sample_60-75_1190546/bams/a8.cov", "Sample_61-76_1190551/bams/a8.cov", "Sample_6-18_1190042/bams/a8.cov", "Sample_62-77_1190560/bams/a8.cov", "Sample_63-78_1190562/bams/a8.cov", "Sample_64-81_1190579/bams/a8.cov", "Sample_65-82_1190580/bams/a8.cov", "Sample_66-83_1190591/bams/a8.cov", "Sample_67-85_1190624/bams/a8.cov", "Sample_68-86_1190629/bams/a8.cov", "Sample_69-87_1190637/bams/a8.cov", "Sample_6a-7_1190827/bams/a8.cov", "Sample_70-88_1190639/bams/a8.cov", "Sample_71-89_1190651/bams/a8.cov", "Sample_7-19_1190044/bams/a8.cov", "Sample_72-90_1190652/bams/a8.cov", "Sample_73-91_1190662/bams/a8.cov", "Sample_74-92_1190670/bams/a8.cov", "Sample_75-93_1190671/bams/a8.cov", "Sample_76-95_1190683/bams/a8.cov", "Sample_77-96_1190685/bams/a8.cov", "Sample_78-97_1190690/bams/a8.cov", "Sample_79-98_1190694/bams/a8.cov", "Sample_7a-9_1190627/bams/a8.cov", "Sample_80-99_1190698/bams/a8.cov", "Sample_81-100_1190700/bams/a8.cov", "Sample_8-20_1190045/bams/a8.cov", "Sample_82-101_1190704/bams/a8.cov", "Sample_83-102_1190705/bams/a8.cov", "Sample_84-103_1190707/bams/a8.cov", "Sample_85-104_1190722/bams/a8.cov", "Sample_86-105_1190731/bams/a8.cov", "Sample_87-107_1190740/bams/a8.cov", "Sample_88-108_1190742/bams/a8.cov", "Sample_89-109_1190746/bams/a8.cov", "Sample_8a-10_1190811/bams/a8.cov", "Sample_90-110_1190747/bams/a8.cov", "Sample_91-111_1190749/bams/a8.cov", "Sample_9-21_1190079/bams/a8.cov", "Sample_92-112_1190753/bams/a8.cov", "Sample_93-113_1190771/bams/a8.cov", "Sample_94-114_1190777/bams/a8.cov", "Sample_95-115_1190784/bams/a8.cov", "Sample_96-116_1190788/bams/a8.cov")


read.vcf = function(x){
    read.table(x, comment.char = "#", sep = "\t", stringsAsFactors = F)
}


file.sizes1 = sapply(all.coverage.files, function(x) file.info(x)$size)
if(length(which(file.sizes1 == 0)) > 0){
	all.coverage.files2 = all.coverage.files[-which(file.sizes1 == 0)]
} else {
	all.coverage.files2 = all.coverage.files
}


# coverage1 = read.delim("Sample_10-23_1190092/bams/a8.cov", sep = "\t", stringsAsFactors = F, header = F)

# for(i in 2:length(all.coverage.files)){	
# 	g = read.delim(all.coverage.files[i], sep = "\t", stringsAsFactors = F, header = F)
# 	num.shared = length(which(coverage1$V2 %in% g$V2)) / nrow(coverage1)
# 	coverage1 = coverage1[which(coverage1$V2 %in% g$V2), ]
	
# 	print(paste0("done comparison with ", all.coverage.files[i]))
# 	print(paste0("% shared is ", num.shared))
# }


coverage1 = read.csv("all.coverage.20x.csv", stringsAsFactors = F)
coverage1 = coverage1[, -1]
coverage1.5 = split(coverage1, coverage1$V1)


all.vcf.files = c("Sample_10-23_1190092/all2.vcf", "Sample_11-24_1190110/all2.vcf", "Sample_12-25_1190126/all2.vcf", "Sample_1-3_1190034/all2.vcf", "Sample_13-26_1190128/all2.vcf", "Sample_14-27_1190139/all2.vcf", "Sample_15-28_1190145/all2.vcf", "Sample_16-29_1190149/all2.vcf", "Sample_17-30_1190160/all2.vcf", "Sample_18-31_1190166/all2.vcf", "Sample_19-32_1190181/all2.vcf", "Sample_1a-1_1190141/all2.vcf", "Sample_20-33_1190199/all2.vcf", "Sample_21-34_1190209/all2.vcf", "Sample_22-35_1190216/all2.vcf", "Sample_23-36_1190218/all2.vcf", "Sample_24-37_1190219/all2.vcf", "Sample_25-38_1190224/all2.vcf", "Sample_26-39_1190224/all2.vcf", "Sample_27-40_1190231/all2.vcf", "Sample_2-8_1190299/all2.vcf", "Sample_28-41_1190239/all2.vcf", "Sample_29-42_1190246/all2.vcf", "Sample_2a-2_1190292/all2.vcf", "Sample_30-43_1190254/all2.vcf", "Sample_3-14_1190007/all2.vcf", "Sample_31-44_1190264/all2.vcf", "Sample_32-45_1190273/all2.vcf", "Sample_33-46_1190281/all2.vcf", "Sample_34-47_1190291/all2.vcf", "Sample_35-48_1190300/all2.vcf", "Sample_36-50_1190313/all2.vcf", "Sample_37-51_1190324/all2.vcf", "Sample_38-52_1190326/all2.vcf", "Sample_39-53_1190349/all2.vcf", "Sample_3a-4_1190238/all2.vcf", "Sample_40-54_1190352/all2.vcf", "Sample_41-55_1190355/all2.vcf", "Sample_4-16_1190032/all2.vcf", "Sample_42-56_1190360/all2.vcf", "Sample_43-57_1190387/all2.vcf", "Sample_44-58_1190396/all2.vcf", "Sample_45-59_1190398/all2.vcf", "Sample_46-60_1190406/all2.vcf", "Sample_47-61_1190420/all2.vcf", "Sample_48-62_1190433/all2.vcf", "Sample_49-63_1190440/all2.vcf", "Sample_4a-5_1190308/all2.vcf", "Sample_50-65_1190451/all2.vcf", "Sample_51-66_1190460/all2.vcf", "Sample_5-17_1190040/all2.vcf", "Sample_52-67_1190468/all2.vcf", "Sample_53-68_1190471/all2.vcf", "Sample_54-69_1190474/all2.vcf", "Sample_55-70_1190475/all2.vcf", "Sample_56-71_1190481/all2.vcf", "Sample_57-72_1190483/all2.vcf", "Sample_58-73_1190496/all2.vcf", "Sample_59-74_1190507/all2.vcf", "Sample_5a-6_1190103/all2.vcf", "Sample_60-75_1190546/all2.vcf", "Sample_61-76_1190551/all2.vcf", "Sample_6-18_1190042/all2.vcf", "Sample_62-77_1190560/all2.vcf", "Sample_63-78_1190562/all2.vcf", "Sample_64-81_1190579/all2.vcf", "Sample_65-82_1190580/all2.vcf", "Sample_66-83_1190591/all2.vcf", "Sample_67-85_1190624/all2.vcf", "Sample_68-86_1190629/all2.vcf", "Sample_69-87_1190637/all2.vcf", "Sample_6a-7_1190827/all2.vcf", "Sample_70-88_1190639/all2.vcf", "Sample_71-89_1190651/all2.vcf", "Sample_7-19_1190044/all2.vcf", "Sample_72-90_1190652/all2.vcf", "Sample_73-91_1190662/all2.vcf", "Sample_74-92_1190670/all2.vcf", "Sample_75-93_1190671/all2.vcf", "Sample_76-95_1190683/all2.vcf", "Sample_77-96_1190685/all2.vcf", "Sample_78-97_1190690/all2.vcf", "Sample_79-98_1190694/all2.vcf", "Sample_7a-9_1190627/all2.vcf", "Sample_80-99_1190698/all2.vcf", "Sample_81-100_1190700/all2.vcf", "Sample_8-20_1190045/all2.vcf", "Sample_82-101_1190704/all2.vcf", "Sample_83-102_1190705/all2.vcf", "Sample_84-103_1190707/all2.vcf", "Sample_85-104_1190722/all2.vcf", "Sample_86-105_1190731/all2.vcf", "Sample_87-107_1190740/all2.vcf", "Sample_88-108_1190742/all2.vcf", "Sample_89-109_1190746/all2.vcf", "Sample_8a-10_1190811/all2.vcf", "Sample_90-110_1190747/all2.vcf", "Sample_91-111_1190749/all2.vcf", "Sample_9-21_1190079/all2.vcf", "Sample_92-112_1190753/all2.vcf", "Sample_93-113_1190771/all2.vcf", "Sample_94-114_1190777/all2.vcf", "Sample_95-115_1190784/all2.vcf", "Sample_96-116_1190788/all2.vcf")


dirs1 = list.dirs(recursive = F)
dirs2 = paste0(dirs1[grep("Sample", dirs1)], "/all_w_indels2.vcf")
all.vcf.files = dirs2

library(dplyr)

#grab only SNP locations that have at least 20x coverage in all samples
# for(i in all.vcf.files){	
	
# 	i2 = gsub("all_w_indels2.vcf", "all_w_indels2.filtered.coverage.vcf", i)        
# 	g = read.delim(i, sep = "\t", comment.char = "#", stringsAsFactors = F, header = F)
	
# 	g2 = split(g, g$V1)
# 	coverage1.5 = split(coverage1, coverage1$V1)

# 	g3 = Map(function(vcf1, cov1){
# 		vcf1[which(vcf1$V2 %in% cov1$V2), ]
# 		}, g2, coverage1.5)

# 	g4 = bind_rows(g3)
	
# 	write.table(g4, i2, sep = "\t", col.names = F, quote = F, row.names = F)
# 	print(paste0("done ", i))
# }



library(dplyr)

all.vcf.files.filtered = paste0(list.files(pattern = "Sample"), "/all_w_indels2.filtered.coverage.vcf") 

all.vcf.files.indel.processed = mclapply(all.vcf.files.filtered, function(x){
	g = read.vcf(x)
	g2 = g[-grep(",", g$V5), ]

	g2.1 = sapply(g2$V4, function(y) nchar(y))
	g2.2 = sapply(g2$V5, function(y) nchar(y))

	deletions1 = which(g2.1 > g2.2 & g2.1 > 1)
	insertions1 = which(g2.1 < g2.2 & g2.2 > 1)

	g.dels = g2[deletions1, ]
	g.ins = g2[insertions1, ]

	#remove insertions and deletions from g2
	g2 = g2[-c(deletions1, insertions1), ]

	g.dels2 = split(g.dels, g.dels$V1) #split by chromosome
	g.dels3 = lapply(g.dels2, function(z){ #split by base position
		split(z, z$V2)
		})
	
	g.dels4 = lapply(g.dels3, function(q){ #change VCF format to include hyphens and only 1 base in REF and ALT columns, here only for deletions
		gg = lapply(q, function(z){						
				bases1 = strsplit(z$V4, "")[[1]]
				bases2 = strsplit(z$V5, "")[[1]]


				z2 = z
				for(i in 1:(length(bases1) - 1)){
					z2 = add_row(z2)
				}
				
				for(i in 1:(length(bases1) - length(bases2))){
					bases2 = c(bases2, "-")
				}

				z2$V4 = bases1
				z2$V5 = bases2

				z2$V2 = z2$V2[[1]]:(z2$V2[[1]] + (length(bases1) - 1))

				z2$V1 = z2$V1[[1]]
				z2	
			})

		bind_rows(gg)

		})

	g.dels5 = bind_rows(g.dels4)

	g.ins2 = split(g.ins, g.ins$V1)
	g.ins3 = lapply(g.ins2, function(z){
		split(z, z$V2)
		})

	g.ins4 = bind_rows(lapply(g.ins3, function(q){ #change VCF format to include hyphens and only 1 base in REF and ALT columns, here only for insertions. 
		bind_rows(lapply(q, function(z){
				bases1 = strsplit(z$V5, "")[[1]]
				bases2 = strsplit(z$V4, "")[[1]]


				z2 = z
				for(i in 1:(length(bases1) - 1)){
					z2 = add_row(z2)
				}
				
				for(i in 1:(length(bases1) - length(bases2))){
					bases2 = c(bases2, "-")
				}

				z2$V5 = bases1
				z2$V4 = bases2

				
				normal.coords1 = which(z2$V4 != "-") 
				insertion.coords1 = which(z2$V4 == "-")
				z2$V2[normal.coords1] = z2$V2[1]:(z2$V2[1] + (length(normal.coords1) - 1))

				insert.coord2 = max(z2$V2[normal.coords1])
				z2$V2[insertion.coords1] = paste0(insert.coord2, "-INS")

				z2$V1 = z2$V1[[1]]
				z2	
			}))
		}))
	
	list(bind_rows(g2, g.dels5), g.ins4)	
}, mc.cores = 30)


all.vcf.w.dels = lapply(all.vcf.files.indel.processed, function(x){
	x[[1]]
	})


all.vcf.ins = lapply(all.vcf.files.indel.processed, function(x){
	x[[2]]
	})

#make dataframes of snp values from VCF files
library(parallel)
#here we produce dataframe containing SNPs from the Watkins VCFs. The dataframe also contains non-SNP loci based on whether we have coverage of these loci
#This means we will have full sequences for downstream phylogenetic analysis which will allow us to avoid the effects of ascertainment bias in phylogenetic inference
snps1 = mclapply(all.vcf.w.dels, function(x){
	#commented code is from non-indel pipeline (only SNPs)	
	# g = read.delim(x, sep = "\t", comment.char = "#", stringsAsFactors = F, header = F)
	# g$V5[which(nchar(g$V5) > 1)] = "-"
	g = x
	g2 = split(g, g$V1)

	#g2 = list of vcf dataframes split by chromosome
	#coverage1.5 = list of coverage dataframes split by chromosome



	coverage.alleles = Map(function(vcf1, cov1){
		cov1$allele = ""		
		cov1$allele[which(cov1$V2 %in% vcf1$V2)] = vcf1$V5	
		cov1	
	}, g2, coverage1.5)


	g3 = bind_rows(coverage.alleles)
	print(paste0("done "))
	g3
}, mc.cores = 20)



all.snps2 = lapply(snps1, function(x){
	colnames(x) = c("chr", "pos", "cov", "allele")
	x
	})


all.snps1 = snps1[[1]]

all.snps3 = do.call(bind_cols, all.snps2)
all.snps4 = all.snps3[, c(1, 2, 3, grep("allele", colnames(all.snps3)))]

library(Biostrings)


iwgsc = readDNAStringSet("~/project.phd.main/genome_assemblies/iwgsc/161010_Chinese_Spring_v1.0_pseudomolecules.fasta") 

all.snps4$ref = ""

iwgsc[["chr1A"]][all.snps4[which(all.snps4$chr == "chr1A"), ]$pos]

all.snps5 = split(all.snps4, all.snps4$chr)
all.snps6 = lapply(all.snps5, function(x){
		x$ref = ""		
		dnaseq1 = strsplit(as.character(iwgsc[[x$chr[1]]][x$pos]), "")
		x$ref = dnaseq1[[1]]
		x
	})

all.snps7 = bind_rows(all.snps6)
#fill in snps from reference sequence
all.snps7[, grep("allele", colnames(all.snps7))] = lapply(all.snps7[, grep("allele", colnames(all.snps7))], function(x){
	x[which(x == "")] = all.snps7$ref[which(x == "")]
	x
	})


# write.csv(all.snps7, "all.snps_w_dels.csv", row.names = F)

sequences1 = lapply(all.snps7[, grep("allele", colnames(all.snps7))], function(x){
	paste0(x, collapse = "")
	})


interleave <- function(v1,v2)                                                                                                                                                                        
{                                                                                                                                                                                                                            
	ord1 <- 2*(1:length(v1))-1
	ord2 <- 2*(1:length(v2))
	c(v1,v2)[order(c(ord1,ord2))]
}                             

seq2 = do.call(c, sequences1) 

all.vcf.files2 = gsub("/all2.vcf", "", all.vcf.files)
all.vcf.files2 = paste0(">", all.vcf.files2)
fasta1 = interleave(all.vcf.files2, seq2)
writeLines(fasta1, "sequences.fa")




# processing cerealsdb data for iqtree - i've now done this in 12062019.pcadapt.R

# cerealsdb1 = read.table("~/project.phd.main/rotation1scripts_v4/original_data/cerealsdb.data/35k.genotypes.vcf", sep = "\t", stringsAsFactors = F)

library(stringr)



#### FURTHER INSERTION PROCESSING #### 

all.vcf.ins

#make the insertions into the same format in which they are in tauschii.blast.processing.R
all.vcf.ins2 = lapply(all.vcf.ins, function(x){	
	x2 = x[grep("INS", x$V2), ]
	x3 = split(x2, x2$V1)
	x4 = bind_rows(lapply(x3, function(z){
		g = split(z, z$V2)
		bind_rows(lapply(g, function(q){			
			q = q[, c(1, 2, 4, 5)]
			q$cs.insert.pos = multi.str.split(q$V2, "-", 1)
			q$insertion = ""
			q$insertion[1] = paste0(q$V5, collapse = "")
			q = q[1, c(1, 5, 6)]
			colnames(q)[1] = "chr"
			q
			}))
		}))
	x4[grep("D", x4$chr), ]
	})

sample.names1 = gsub("/all_w_indels2\\.filtered\\.coverage\\.vcf", "", all.vcf.files.filtered)

all.vcf.ins3 = Map(function(x, y){
	x$var = y
	x
	}, all.vcf.ins2, sample.names1)

all.vcf.ins4 = bind_rows(all.vcf.ins3)
# s(all.vcf.ins4, "~/project.phd.main/rotation1scripts_v4/saved.objects/all.vcf.ins4", "processing.watkins.vcfs.R")

